plugins/upower.c

```diff
@@ -0,0 +1,478 @@
+}
+
+static void emulator_hfp_watch(struct ofono_atom *atom,
+				enum ofono_atom_watch_condition cond,
+				void *data)
+{
+	struct ofono_modem *modem = data;
+
+	if (cond == OFONO_ATOM_WATCH_CONDITION_REGISTERED) {
+		DBusMessageIter iter;
+		DBusPendingCall *call;
+		DBusMessage *msg;
+		const char *iface = UPOWER_DEVICE_INTERFACE;
+		const char *property = "Percentage";
+
+		DBG("%s: REGISTERED", __func__);
+
+		/* TODO: handle removable batteries */
+
+		modems = g_list_append(modems, modem);
+
+		if (modems->next != NULL)
+			return;
+
+		/* Query current battery value */
+		msg = dbus_message_new_method_call(UPOWER_SERVICE,
+						battery_device_path,
+						DBUS_PROPERTIES_INTERFACE,
+						"Get");
+		if (msg == NULL) {
+			ofono_error("%s: dbus_message_new_method failed",
+					__func__);
+			return;
+		}
+
+		dbus_message_iter_init_append(msg, &iter);
+		dbus_message_iter_append_basic(&iter,
+						DBUS_TYPE_STRING, &iface);
+		dbus_message_iter_append_basic(&iter,
+						DBUS_TYPE_STRING, &property);
+
+		if (!dbus_connection_send_with_reply(connection, msg,
+							&call, -1)) {
+			ofono_error("%s: Sending EnumerateDevices failed",
+					__func__);
+			goto done;
+		}
+
+		dbus_pending_call_set_notify(call, get_property_reply, NULL, NULL);
+done:
+		dbus_message_unref(msg);
+
+	} else {
+		DBG("%s: UNREGISTERED", __func__);
+
+		modems = g_list_remove(modems, modem);
+		if (modems != NULL)
+			return;
+
+		g_dbus_remove_watch(connection, upower_battery_watch);
+	}
+}
+
+static void modemwatch(struct ofono_modem *modem, gboolean added, void *user)
+{
+
+	const char *path = ofono_modem_get_path(modem);
+
+	DBG("modem: %s, added: %d", path, added);
+
+	if (added) {
+		guint watch_id;
+
+		watch_id = __ofono_modem_add_atom_watch(modem,
+					OFONO_ATOM_TYPE_EMULATOR_HFP,
+					emulator_hfp_watch, modem, NULL);
+
+		g_hash_table_insert(modem_hfp_watches, g_strdup(path),
+						GINT_TO_POINTER(watch_id));
+
+	} else {
+		guint *watch_id = g_hash_table_lookup(modem_hfp_watches, path);
+
+		if (watch_id != NULL) {
+			__ofono_modem_remove_atom_watch(modem, *watch_id);
+			g_hash_table_remove(modem_hfp_watches, path);
+		}
+	}
+}
+
+static void call_modemwatch(struct ofono_modem *modem, void *user)
+{
+	modemwatch(modem, TRUE, user);
+}
+
+static gboolean parse_devices_reply(DBusMessage *reply)
+{
+	DBusMessageIter array, iter;
+	GPatternSpec *match_pattern;
+	const char *path;
+
+	DBG("");
+
+	if (dbus_message_get_type(reply) == DBUS_MESSAGE_TYPE_ERROR) {
+		ofono_error("%s: ERROR reply to EnumerateDevices", __func__);
+		return FALSE;
+	}
+
+	if (dbus_message_iter_init(reply, &array) == FALSE) {
+		ofono_error("%s: error initializing array iter", __func__);
+		return FALSE;
+	}
+
+	if (dbus_message_iter_get_arg_type(&array) != DBUS_TYPE_ARRAY) {
+		ofono_error("%s: type != ARRAY!", __func__);
+		return FALSE;
+	}
+
+	dbus_message_iter_recurse(&array, &iter);
+
+	match_pattern = g_pattern_spec_new("*/battery_*");
```

Comment:
I don't really see the point of using a pattern as this would be equivalent to finding a substring using strstr()